cast/src/lib.rs

```diff
@@ -1656,6 +1661,63 @@ impl SimpleCast {
     pub fn disassemble(bytecode: &str) -> Result<String> {
         format_operations(disassemble_str(bytecode)?)
     }
+
+    /// Gets the selector for a given function signature
+    /// Optimizes if the `optimize` parameter is set to a number of leading zeroes
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use cast::SimpleCast as Cast;
+    ///
+    /// fn main() -> eyre::Result<()> {
+    ///     assert_eq!(Cast::get_selector("foo(address,uint256)", None)?.0, String::from("0xbd0d639f"));
+    ///
+    ///     Ok(())
+    /// }
+    /// ```    
+    pub fn get_selector(signature: &str, optimize: Option<usize>) -> Result<(String, String)> {
+        let optimize = optimize.unwrap_or(0);
+        if optimize > 4 {
+            eyre::bail!("Number of leading zeroes must not be greater than 4");
+        }
+        if optimize == 0 {
+            let selector = HumanReadableParser::parse_function(signature)?.short_signature();
+            return Ok((format!("0x{}", hex::encode(selector)), String::from(signature)))
+        }
+        let Some((name, params)) = signature.split_once('(') else {
+            eyre::bail!("Invalid signature");
+        };
+
+        let num_threads = num_cpus::get();
+        let found = AtomicBool::new(false);
+
+        let result: Option<(u32, String, String)> =
+            (0..num_threads).into_par_iter().find_map_any(|i| {
+                let nonce_start = i as u32;
+                let nonce_step = num_threads as u32;
+
+                let mut nonce = nonce_start;
+                while nonce < std::u32::MAX && !found.load(Ordering::Relaxed) {
```

Comment:
ah okay, this spawns num_threads worker threads that use the num_threads as step to loop over all nonces.

this is probably more efficient than iterating over 0..u32::Max range
but requires the atomic bool

Reply:
It is indeed, like 2-3x faster than `(0..).par_bridge` last time i checked it
https://github.com/foundry-rs/foundry/pull/4705#discussion_r1161328181

Reply:
Are you sure it needs `Arc<AtomicBool>`? `AtomicBool` isn't `Copy`

Reply:
ah yeh you're right of course, the `find_map_any` doesn't require 'static so the bool isn't actually moved
confused this with `spawn`